Add manual review page in External Security Reviews framework.

the-caliber opened the pull request:
**Changes:**
Hi, as discussed previously on TG, i have created a PR, I created another page as the mental model was more inclined towards the approach for “manual review”. 

Please let me know your thoughts on this @PatrickAlphaC 
If this dedicated page makes sense And if any changes are needed!

Thank you!

PatrickAlphaC commented:
Hmm... This is a bit of a re-skin of what is in `external-security-reviews`. I think it would be better to have one canonical place for this type of review, and we can link to it, that way we don't have to update both folders/files.

Also, the formatting is very off in your markdown, could you please follow the new docs from this PR to format your markdown?

https://github.com/security-alliance/frameworks/pull/171

the-caliber commented:
Hey there!

Yes, i noticed the `external-security-reviews` is there. was wondering where to add this info/page(manual review).

1. Do you want me to add `manual-review.md` to `external-security-reviews/` folder ? so that it can be linked to the `Manual Review:` in the [src/external-security-reviews/smart-contracts/README.md](https://github.com/security-alliance/frameworks/blob/develop/src/external-security-reviews/smart-contracts/README.md#audit-methodologies) ? Can you expand more on it :)

2. Also could you give more idea about the reason behind keeping `external-security-reviews` section different ?
Maybe you already gave it here:
>I think it would be better to have one canonical place for this type of review, and we can link to it

3. About `formatting` , Not sure why this seems **off** (at least when not rendered). But sure, i will look into your recommendation about #171

Thanks!

the-caliber commented:
Yep! noticed its now 184.

---
Just waiting for @PatrickAlphaC to comment on my questions (especially 1st) and will fix the linting etc...

PatrickAlphaC commented:
I like your suggestion for your point 1 @the-caliber! 

2. We could just have a section called `security reviews` and have the difference between internal and external? It just feels like a pretty big violation of DRY (hehe) if we put the same information about manual reviews in 2 places. Internal or external, manual reviews are the same. Maybe we just use the `manual reviews` section of the `external security reviews`. What do you think? 

3. Formatting PR is merged now!

the-caliber commented:
Hey @mattaereal , If you are seeing this, 

To which md file the outline should be updated when making PR?

the [contributing doc](https://frameworks.securityalliance.dev/contribute/contributing.html#handling-the-summary) and PR template suggest that `If you're modifying the general outline, make sure to use src/config/SUMMARY.develop and not src/SUMMARY.md`

While locally hosting the mdbook, making changes to `SUMMARY.md.develop` does not reflect in the hosted mdbook. but when i make changes to the `SUMMARY.md` it reflects the changes (changes about outline/index) in the hosted mdbook.

the-caliber commented:
Hi @PatrickAlphaC Now it should reflect the changes in this vercel preview deployments.

I have edited `manual-review.md` to take care of formatting and spellings with markdownlint-cli2 and aspell respectively.
(Not sure why the file looks off when not rendered. I think its because i am using code snippet in the list item. But when rendered/previewed it seems good.)

Moved `manual-review.md` to the 'external-security-reviews/smart-contracts'
Also updated the `SUMMARY.md.develop` to reflect the security testing correct outline.

I understand that the `Automated review` can't be in the manual review so i have just added references to it. I thought about adding a note similar to "automated review is not part of manual review" but generally we are assuming the viewer already know that. i think removing the `Automated review` can also make sense. let me know your thoughts.

mattaereal commented:
Added a few suggestions! I accidentally broke the format by suggesting something that did not allow me to complete through the web ui 😓 .

Also, remember the PR template asks you to modify `src/config/SUMMARY.develop`, not SUMMARY.md

the-caliber commented:
Hi,

Can you point out what are the suggestions you added ? Currently unable to see them in this thread. Are these in the PR I am making or somewhere else.

Also noted that there's a conflict, is it related to the thing you are saying `accidentally broke the format by suggesting something that did not allow me to complete through the web ui` ?

---
> Also, remember the PR template asks you to modify src/config/SUMMARY.develop, not SUMMARY.md  

Yes, I remember this. In current PR i made changes in both SUMMARY.md.develop and SUMMARY.md because the SUMMARY.md  was behind (I know it should be only used for local testing/hosting mdbook as you said). So I thought it is a good practice to keep it updated because when others will pull the repo to locally test it they will be confused by seeing differences in the outline.

Please let me know if keeping the SUMMARY.md.develop updated (and adding it in PR) creates any problems.

mattaereal reviewed (changes requested):
Some of the requested changes, but pay attention to my last comments as well :).

the-caliber commented:
I will resolve the suggestions soon.  Now I am able to see the suggestions.

the-caliber commented:
> > Hi,
> > Can you point out what are the suggestions you added ? Currently unable to see them in this thread. Are these in the PR I am making or somewhere else.
> > ~Also noted that there's a conflict, is it related to the thing you are saying `accidentally broke the format by suggesting something that did not allow me to complete through the web ui` ?~ Seems, no.
> > **PS:** saw the discord messages, You are talking about `add mermaid syntax and replace your images of several type of graphs for mermaid compatible ones`. Still, wondering where you made them(suggestions).
> > > Also, remember the PR template asks you to modify src/config/SUMMARY.develop, not SUMMARY.md
> > 
> > 
> > Yes, I remember this. In current PR i made changes in both SUMMARY.md.develop and SUMMARY.md because the SUMMARY.md was behind (I know it should be only used for local testing/hosting mdbook as you said). So I thought it is a good practice to keep it updated because when others will pull the repo to locally test it they will be confused by seeing differences in the outline.
> > Please let me know if keeping the SUMMARY.md.develop updated (and adding it in PR) creates any problems.
> 
> Huh, this is weird, it is native functionality of GitHub. Send me a screenshot of the notifications of this thread please. You should be able to see them all above my last comment, and even get notifications from them.
> 
> Here's a direct link to the files where you should also see them as well: https://github.com/security-alliance/frameworks/pull/183/files/4b3eedfde8bbbc692160f8eed4d828612c26ec4c

### **Here's the screenshot as you asked:**
<img width="631" height="947" alt="image" src="https://github.com/user-attachments/assets/92c2559d-4afc-4ad1-bc73-7cd60e72d4fd" />


### **At that point i couldn't see suggestions.**

### **Currently i can see after your last message:**
<img width="751" height="948" alt="image" src="https://github.com/user-attachments/assets/cd2c0bb3-c250-45ad-a74b-f6633945fca1" />

Still, let me know if that was any mistake from my side :)

the-caliber commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,234 @@
+
+# Manual Review
+
+Manual review of a smart contract is the process where a security engineer carefully reads through the source code to identify potential vulnerabilities, logic errors, and design flaws.
```
What about this ? `Manual review of a smart contract is the process of carefully examining the source code to identify potential vulnerabilities, logic errors, and design flaws.`

the-caliber commented:
I have updated it.

1. Used Mermaid diagram.
2. Changed the **Manual Review** definition.
3. Added descriptions for all the points. (except `Functional testing` for now. Suggestions welcome)

Please let me know if i am missing something.

mattaereal reviewed (approved):
Thanks for this! Seems to be ok from my side. Waiting on @Robert-MacWha that is also taking a last look.

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+
+## Report writing  
+
+1. **Doubts and issue discussion with internal team:**  
```
```suggestion
1. **Questions and issue discussion with internal team:**  
```

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+1. **Doubts and issue discussion with internal team:**  
+Discuss any doubts and identified issues with the internal team or, if applicable, with the team of reviewers working in parallel on the project. This will help you reach a conclusion.  
+
+2. **Doubts and issue discussion with the developers or project team:**  
```
```suggestion
2. **Questions and issue discussion with the developers or project team:**  
```

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
````diff
@@ -0,0 +1,288 @@
+    ```
+
+7. **Write down notes, doubts and edge cases:**  
+One reason for writing down doubts, ideas, possible issues is, when you start going through these issues as you encounter new path you can see more doubts, possible issues in your mind. so sometimes it becomes a loop where you will keep thinking about new ideas while exploring previous/current one.  
````
Replacing the use of "doubts" with questions, since I think that's more accurate to the meaning you want to convey.  "doubts" to me implies that you doubt your understanding, whereas questions are questions about the code.

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
By fixed code, do you mean like code fixed to a specific commit has / release?  Or something else

the-caliber replied:
By `Fixed code review` I mean. The updated code received from the developers after the reviewer provides initial audit report to them.

mattaereal replied:
Yeah I agree, this is actually called a recheck btw. You could go with: "Recheck - updated code review" or something explicit like that.

mattaereal replied:
Nah, it's ok!

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```suggestion
For checking fixed code, it may be necessary to repeat previously performed actions, depending on what has changed. To avoid repeating work, you can:
```

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```suggestion
This helps in identifying updates and detecting deviations from expected changes. For example, a git diff may reveal changes in functions that should not have been changed.
```

I think this is what you mean by visual level?  "Oh no - the onlyAdmin function shouldn't show a diff"

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
This feels like the same as point 1 - is it?  If so, might as well remove it / combine the two

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+3. **Documentation and specification:**  
+    - Documentation describes how the code works, how to use it, and its expected functionality.
+    - Specification defines the exact requirements, constraints, and behaviors the system must meet, including invariants that must always hold true.
+        - Invariant is the property of the system that should always hold. Invariants can be useful for checking the expected behavior of the system when the certain actions or a transaction flow is executed. Invariants are useful in many testing approaches, including manual reviews, where a reviewer checks whether a specific property can be violated for a function under any sequence of function  ...
```
```suggestion
        - Invariants are the properties of the system that should always hold. Invariants can be useful for checking the expected behavior of the system when the certain actions or a transaction flow is executed. Invariants are useful in many testing approaches, including manual reviews, where a reviewer checks whether a specific property can be violated for a function under any sequence of function calls.
```

Robert-MacWha commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+    - Documentation describes how the code works, how to use it, and its expected functionality.
+    - Specification defines the exact requirements, constraints, and behaviors the system must meet, including invariants that must always hold true.
+        - Invariant is the property of the system that should always hold. Invariants can be useful for checking the expected behavior of the system when the certain actions or a transaction flow is executed. Invariants are useful in many testing approaches, including manual reviews, where a reviewer checks whether a specific property can be violated for a function under any sequence of function  ...
+        - Get simplified math formulas if the protocol has any custom/fancy formulas implemented.
```
Just a though, but do we want to add like a "tips!" section at the bottom / top for these?  Since it's a good point to make, and will be helpful, but also doesn't really fit here?

I don't feel strongly about this, was just an option.

the-caliber replied:
Okay.

By this:
> It's relevant enough that it fits here

I assume you also find this point relevant in the `Documentation and specification`. So I kept it as it is for now.

Still let me know if i misunderstood and you wanted to say something else.

Robert-MacWha replied:
You understood correctly - seems good!

mattaereal commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+
+4. **Tests:**  
+Tests are important for understanding the expected behavior and ensuring that the system works as intended. The following types of tests can be useful for understanding and verifying the code:  
+    - [unit tests](../../security-testing/unit-testing.md): Unit tests ensure that individual functions behave correctly under expected conditions.
```
Oh, I completely missed that on Patrick's merge! Yes, can you add them both to those files please? It will fix this as robert is mentioning.

the-caliber replied:
Done. updated both `Summary.md.develop` and `Summary.md.main`

Also for SUMMARY.md

> SUMMARY.md was behind (I know it should be only used for local testing/hosting mdbook as you said). So I thought it is a good practice to keep it updated because when others will pull the repo to locally test it they will be confused by seeing differences in the outline.
> 
> Please let me know if keeping the SUMMARY.md.develop updated (and adding it in PR) creates any problems.

Robert-MacWha replied:
Good call - keeping it updated's probably best.

mattaereal commented:
Make sure to resolve all conversations after reaching consensus!

the-caliber commented on src/external-security-reviews/smart-contracts/manual-review.md:
```diff
@@ -0,0 +1,288 @@
+  This can be done by looking into a diagram and current understanding of the smart contract functionality. For example in this case it's a Token contract that is least dependent ( in fact the functionality from that contract is getting used in other contracts so other smart contracts are dependent on it).
+
+3. **Checklist reviews:**  
+    - Known smart contracts vulnerabilities.
```
Added. Since the SWC is not maintained I mentioned some others. Suggestions are welcome. Changes should look like:

3. **Checklist reviews:**  
    - Known smart contracts vulnerabilities.
    - Project specific checklists (e.g. When dealing with bridges it's also important to check bridge specific checklists).
    - Checklists may vary over time and can become outdated. Below are some examples for reference:
        - [Smart Contract Security Field Guide](https://scsfg.io/hackers/)
        - [Smart Contract Security Verification Standard](https://github.com/ComposableSecurity/SCSVS?tab=readme-ov-file#table-of-contents)

Robert-MacWha replied:
yep, those seem good!  It's just nice to link to external resources - give people more information.

the-caliber replied:
Yes makes sense! 
Resolving this!

the-caliber commented on src/external-security-reviews/smart-contracts/manual-review.md:
````diff
@@ -0,0 +1,288 @@
+    }
+    ```  
+
+    In next example, because every value (`_amount`, `minRequired`) can go above and below, the results can vary. If the result is used for some operations like decreasing the `drivingScore` in `burnFuelAndReduce()` for example.
````
Used `increase or decrease` instead of `above and below`

Robert-MacWha replied:
Sorry!  Didn't notice this one.

Yep, `can increase or decrease` makes more sense.  Or `varies` - but whichever works.

the-caliber commented:
I have resolved the open conversations. still, there are 3 open conversations. I have commented on them, once i get your comment/reaction, i will resolve them as well.

Feel free to reopen any resolve conversation if you think!

mattaereal commented:
Ident everything properly for readability please :)

Robert-MacWha commented:
> One important question about usage of indentation for the list item's code snippets:
> 
> As you can see I am using most of the code blocks (solidity, mermaid, minimal md file) as a part of numbered list items, so most of them are **indented**. Now with some recent modification from Robert the `The minimal markdown file for taking notes might look like this:` is out of the indentation, so is the actual markdown code snippet.
> 
> Should I still keep using indentation for other snippets or they should be taken out of indentation ?? (My views are , Out of indentation looks good (since it gets more space), but it then creates visual difference where the code snippet looks little out of the list item.)
> 
> Both works for me, Maybe we just need to follow one format!

I was just making suggestions from within github's review UI lol - it doesn't do a great job at preserving style.  Sorry if I messed up something by accident!
